### PR TITLE
Make SelectionCache and EvaluationCache thread-safe

### PR DESCRIPTION
**Pull request by Zoxc**
Split out from https://github.com/rust-lang/rust/pull/49558

r? @nikomatsakis

### CODE REVIEW

**Comment by TimNN** [comment]
Your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/364512798) ([raw log](https://api.travis-ci.org/v3/job/364512798/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
Resolving deltas: 100% (613496/613496), completed with 4872 local objects.
---
[00:00:41] configure: rust.quiet-tests     := True
---
[00:37:20] thread 'main' panicked at 'assertion failed: *old == value', librustc_data_structures/sync.rs:236:42
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

File: src/librustc/traits/select.rs [hunk 1]
```diff
@@ -1296,15 +1298,17 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
...
                     debug!(
                         "insert_candidate_cache(trait_ref={:?}, candidate={:?}) global",
                         trait_ref,
                         candidate,
                     );
-                    cache.insert(trait_ref, WithDepNode::new(dep_node, candidate));
+                    // This may overwrite the cache with the same value
+                    tcx.selection_cache
+                       .hashmap.borrow_mut()
+                       .insert(trait_ref, WithDepNode::new(dep_node, candidate));
```
Comment (nikic) [inline src/librustc/traits/select.rs]: Given the comment above, should this also use `insert_same()`?

Comment (Zoxc) [reply src/librustc/traits/select.rs]: The inserted type doesn't implement `Eq`.

**Review by nikomatsakis — APPROVED** [review]

**Comment by Zoxc** [comment]
This doesn't work since `insert_evaluation_cache` overwrites values.

@bors r-

**Comment by nikomatsakis** [comment]
@Zoxc to be honest I didn't think too hard here-- and I still haven't. =) Can you spell out the problem, and do you have any ideas for how to solve?

**Comment by Zoxc** [comment]
@nikomatsakis I though maps in `SelectionCache` and `EvaluationCache` would have values which are deterministically based on their keys. That is not true for `EvaluationCache` at least, since it will change the value of a`ty::PolyTraitRef<'tcx>` key to a different `WithDepNode<EvaluationResult>`. I'm wondering if this is expected behavior, and if so, how to make these caches thread-safe?

**Comment by shepmaster** [comment]
Ping from triage, @Zoxc and @nikomatsakis ! Will you have time to make progress on this in the near future?

**Comment by shepmaster** [comment]
Ping from triage, @Zoxc ! You have test failures.

**Comment by Zoxc** [comment]
I've created https://github.com/rust-lang/rust/issues/50507 for the bug which overwrites the cache and added a fixme to use `insert_same` when that is fixed.

**Review by nikomatsakis — APPROVED** [review]

**Pull request merged**